### ISSUE TITLE
Enable more concise configuration of top-level properties that abstract component details

### ISSUE THREAD

**Issue by magistrula** [san650/ember-cli-page-object#295]
I really like organizing page objects in terms of components because I think doing so helps to make page objects coherent and well-organized, but it can lead to tests that are tightly coupled to the page structure.

Instead of writing tests that look like this:
```javascript
it('disables the Roster form if the user inputs the name "Voldemort"', function() {
  RosterPageObject.form.input.fillIn('Voldemort');
  RosterPageObject.form.submitButton.click();

  expect(RosterPageObject.form.input.isDisabled).to.be.true;
});
```
I'd like to write tests that look like this:
```javascript
it('disables the Roster form if the user inputs the name "Voldemort"', function() {
  RosterPageObject.addName('Voldemort');

  expect(RosterPageObject.isNameInputDisabled).to.be.true;
});
```
Currently, this is how I'd set up my page objects if I want to write tests as shown in the second snippet above:
```javascript
// Imagine that MyForm is a generic form component, 
// actually defined in its own page object file.
const MyForm = {
  scope: '.my-form',

  input: {
    scope: 'input',
    isDisabled: is(':disabled')
  },

  submitButton: {
    scope: 'button'
  }
};

// This is the page object for the component we're going to test
const RosterPageObject = create({
  scope: '.roster',

  form: MyForm,
  
  addName(name) {
    this.form.input.fillIn(name);
    this.form.submitButton.click();
  },

  isNameInputDisabled: {
    isDescriptor: true,
    get() {
      return this.form.input.isDisabled;
    }
  }
});
```
It would be nice to have some helpers (e.g., `component`, `pipe` and `alias` in the examples below) that make setting up these top-level properties easier and more concise.
```javascript
const MyForm = {
  scope: '.my-form',

  input: component('input', {
    isDisabled: is(':disabled')
  },

  submitButton: component('button')
};

const RosterPageObject = create({
  scope: '.roster',

  form: MyForm,
  
  addName: pipe('form.input.fillIn', 'form.submitButton.click'),
  isNameInputDisabled: alias('form.input.isDisabled')
});
```
The helpers I have in mind so far are: 
- `alias`: sets up a property that returns the value of some other property
- `component`: sugar for setting up a POJO with a `scope` property
- `conditional-alias`: sets up a property that can return various different values depending on some set of conditions to be checked each time this property is accessed
- `ternary-alias`: a simpler version of `conditional-alias`, with only one condition to check
- `descriptor`: sugar for `{ isDescriptor: true, get() { ... } }`
- `every`: checks each property passed to it and returns `true` if all of those properties are truthy. If any of the properties is a collection, a collection with count > 1 will be considered truthy.
- `none`: checks each property passed to it and returns `true` if all of those properties are falsy. If any of the properties is a collection, a collection with count === 0 will be considered falsy.

**Comment by san650** [comment]
@magistrula I really like the `pipe` and `alias` operators, I think they could be a really nice addition to the add-on.

I would love to understand a bit more how the `pipe` operator would behave, in your previous example I see that the function returned by `pipe` receives the argument for `form.input.fillIn`... What would be the exact behavior for this operator? What will happen if you use actions and non-action properties inside a `pipe` definition? `foo: alias('form.input.fillIn', 'form.submitButton.click', 'form.input.value') `form.input.value` is a sync call whereas `form.submitButton.click` is an async one.

About the `component` helper, I think it's mostly syntactic sugar, we're just replacing `{ scope: "foo" }` by a function helper... I think the name of the helper could be a bit confusing for new users because we already have the `page object component` concept that it's a bit different than this helper.

If you are interested in develop these operators (pipe and alias) please open separate PR for each one, and discuss a bit more how `pipe` operator will behave.

**Comment by simonihmig** [comment]
Allow me to chime in here... 

I very much like the idea of the `alias` helper. We use the page object components most of the time also for all component integration tests (that's were the need of the component-test blueprint I contributed a while ago came from). But when reusing these components for the top level page objects within acceptance tests, we end up in a similar situation, having the nesting of components "visible" in our tests by the similar nesting of page object components (as in your `RosterPageObject.form.input...` example). When refactoring the UI, maybe moving a nested component to some other place in the component tree, these changes also break the tests, as the nesting is hardcoded in our actual test files, rather than just within the page object, which kind of defies the purpose of this addon, abstracting the UI/DOM structure away in tests.

So 👍  for `alias`!

**Comment by san650** [comment]
@magistrula I think this issue has grown too much and is difficult to discuss over each property separately. There are some of the suggested properties that could be added without too much discussion (`every`, `none`) and some of them that need more discussions and examples of what they do and how they behave, e.g. How do you use `conditional-alias`?

What do you think? If it's too much work, let me know and I can help with the creation of the issues (although for some of them I don't fully understand what they do)

**Comment by magistrula** [comment]
Sounds good, @san650; I'll open separate issues for each proposed helper.

**Comment by san650** [comment]
Now that `alias` is merged (see #296), I'll close this issue in favor of individual issues for each helper.
